clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java

```diff
@@ -922,6 +946,103 @@ public void testTransactionManagerEnablesV2() {
         assertTrue(transactionManager.isTransactionV2Enabled());
     }
 
+    @Test
+    public void testTransactionV2AddPartitionAndOffsets() throws InterruptedException {
+        initializeTransactionManager(Optional.of(transactionalId), true);
+        doInitTransactions();
+
+        transactionManager.beginTransaction();
+
+        Future<RecordMetadata> responseFuture = appendToAccumulator(tp0);
+
+        assertFalse(responseFuture.isDone());
+
+        prepareProduceResponse(Errors.NONE, producerId, epoch);
+        assertTrue(transactionManager.transactionContainsPartition(tp0));
```

Comment:
just for my understanding, the partition will be added before the produce response returns?

Reply:
Updated this part. And yes, the partition is added before the response.